File: graph.java
import java.util.*;

public class graph {
    private Map<String, List<String>> adjList = new HashMap<>();

    public void addVertex(String v) {
        if (!adjList.containsKey(v)) {
            adjList.put(v, new ArrayList<>());
            System.out.println("Vertex '" + v + "' ditambah.");
        } else {
            System.out.println("Vertex sudah wujud.");
        }
    }

    public void removeVertex(String v) {
        if (adjList.containsKey(v)) {
            for (String neighbor : adjList.get(v)) {
                adjList.get(neighbor).remove(v);
            }
            adjList.remove(v);
            System.out.println("Vertex '" + v + "' dihapus.");
        } else {
            System.out.println("Vertex tidak ditemui.");
        }
    }

    public void addEdge(String v1, String v2) {
        if (adjList.containsKey(v1) && adjList.containsKey(v2)) {
            if (!adjList.get(v1).contains(v2)) {
                adjList.get(v1).add(v2);
                adjList.get(v2).add(v1);
                System.out.println("Edge antara '" + v1 + "' dan '" + v2 + "' ditambah.");
            } else {
                System.out.println("Edge sudah wujud.");
            }
        } else {
            System.out.println("Salah satu atau kedua-dua vertex tidak ditemui.");
        }
    }

    public void removeEdge(String v1, String v2) {
        if (adjList.containsKey(v1) && adjList.containsKey(v2)) {
            if (adjList.get(v1).contains(v2)) {
                adjList.get(v1).remove(v2);
                adjList.get(v2).remove(v1);
                System.out.println("Edge antara '" + v1 + "' dan '" + v2 + "' dihapus.");
            } else {
                System.out.println("Edge tidak ditemui.");
            }
        } else {
            System.out.println("Vertex tidak ditemui.");
        }
    }

    public void displayMatrix() {
        if (adjList.isEmpty()) {
            System.out.println("Graph kosong.");
            return;
        }

        List<String> vertices = new ArrayList<>(adjList.keySet());
        Collections.sort(vertices);

        System.out.println("\nAdjacency Matrix:");
        System.out.print("  ");
        for (String v : vertices) {
            System.out.print(v + " ");
        }
        System.out.println();

        for (String v1 : vertices) {
            System.out.print(v1 + " ");
            for (String v2 : vertices) {
                if (adjList.get(v1).contains(v2)) {
                    System.out.print("1 ");
                } else {
                    System.out.print("0 ");
                }
            }
            System.out.println();
        }
    }

    public void dfs(String start) {
        if (!adjList.containsKey(start)) {
            System.out.println("Vertex awal tidak ditemui.");
            return;
        }
        System.out.print("Traversal DFS: ");
        Set<String> visited = new HashSet<>();
        dfsUtil(start, visited);
        System.out.println();
    }

    private void dfsUtil(String vertex, Set<String> visited) {
        visited.add(vertex);
        System.out.print(vertex + " ");

        List<String> neighbors = new ArrayList<>(adjList.get(vertex));
        Collections.sort(neighbors);
        
        for (String neighbor : neighbors) {
            if (!visited.contains(neighbor)) {
                dfsUtil(neighbor, visited);
            }
        }
    }

    public void bfs(String start) {
        if (!adjList.containsKey(start)) {
            System.out.println("Vertex awal tidak ditemui.");
            return;
        }

        Set<String> visited = new HashSet<>();
        Queue<String> queue = new LinkedList<>();

        visited.add(start);
        queue.add(start);

        System.out.print("Traversal BFS: ");
        while (!queue.isEmpty()) {
            String vertex = queue.poll();
            System.out.print(vertex + " ");

            List<String> neighbors = new ArrayList<>(adjList.get(vertex));
            Collections.sort(neighbors);

            for (String neighbor : neighbors) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.add(neighbor);
                }
            }
        }
        System.out.println();
    }

    public static void main(String[] args) {
        graph g = new graph();
        Scanner scanner = new Scanner(System.in);
        int pilihan;

        do {
            System.out.println("\nMENU:");
            System.out.println("1. Tambah Vertex");
            System.out.println("2. Hapus Vertex");
            System.out.println("3. Tambah Edge");
            System.out.println("4. Hapus Edge");
            System.out.println("5. Tampilkan graph (Matrix)");
            System.out.println("6. Traversal DFS");
            System.out.println("7. Traversal BFS");
            System.out.println("8. Quit");
            System.out.print("Pilih menu (1-8): ");
            
            pilihan = scanner.nextInt();
            scanner.nextLine(); 

            switch (pilihan) {
                case 1:
                    System.out.print("Masukkan nama vertex: ");
                    g.addVertex(scanner.nextLine());
                    break;
                case 2:
                    System.out.print("Masukkan nama vertex untuk dihapus: ");
                    g.removeVertex(scanner.nextLine());
                    break;
                case 3:
                    System.out.print("Masukkan vertex 1: ");
                    String v1 = scanner.nextLine();
                    System.out.print("Masukkan vertex 2: ");
                    String v2 = scanner.nextLine();
                    g.addEdge(v1, v2);
                    break;
                case 4:
                    System.out.print("Masukkan vertex 1: ");
                    String e1 = scanner.nextLine();
                    System.out.print("Masukkan vertex 2: ");
                    String e2 = scanner.nextLine();
                    g.removeEdge(e1, e2);
                    break;
                case 5:
                    g.displayMatrix();
                    break;
                case 6:
                    System.out.print("Masukkan vertex awal untuk DFS: ");
                    g.dfs(scanner.nextLine());
                    break;
                case 7:
                    System.out.print("Masukkan vertex awal untuk BFS: ");
                    g.bfs(scanner.nextLine());
                    break;
                case 8:
                    System.out.println("Program tamat.");
                    break;
                default:
                    System.out.println("Pilihan tidak sah.");
            }
        } while (pilihan != 8);
        
        scanner.close();
    }
}

File: graph.py
from collections import deque

class Graph:
    def __init__(self):
        self.graph = {}

    def add_vertex(self, v):
        if v not in self.graph:
            self.graph[v] = []
            print(f"Vertex '{v}' ditambah.")
        else:
            print("Vertex sudah wujud.")

    def remove_vertex(self, v):
        if v in self.graph:
            for neighbor in self.graph[v]:
                self.graph[neighbor].remove(v)
            del self.graph[v]
            print(f"Vertex '{v}' dihapus.")
        else:
            print("Vertex tidak ditemui.")

    def add_edge(self, v1, v2):
        if v1 in self.graph and v2 in self.graph:
            if v2 not in self.graph[v1]:
                self.graph[v1].append(v2)
                self.graph[v2].append(v1)
                print(f"Edge antara '{v1}' dan '{v2}' ditambah.")
            else:
                print("Edge sudah wujud.")
        else:
            print("Salah satu atau kedua-dua vertex tidak ditemui.")

    def remove_edge(self, v1, v2):
        if v1 in self.graph and v2 in self.graph:
            if v2 in self.graph[v1]:
                self.graph[v1].remove(v2)
                self.graph[v2].remove(v1)
                print(f"Edge antara '{v1}' dan '{v2}' dihapus.")
            else:
                print("Edge tidak ditemui.")
        else:
            print("Vertex tidak ditemui.")

    def display_matrix(self):
        vertices = sorted(list(self.graph.keys()))
        if not vertices:
            print("Graph kosong.")
            return
        
        print("\nAdjacency Matrix:")
        print("  " + " ".join(vertices))
        for v1 in vertices:
            row = [v1]
            for v2 in vertices:
                if v2 in self.graph[v1]:
                    row.append("1")
                else:
                    row.append("0")
            print(" ".join(row))

    def dfs(self, start, visited=None, is_first=True):
        if start not in self.graph:
            print("Vertex awal tidak ditemui.")
            return
        if visited is None:
            visited = set()
            print("Traversal DFS:", end=" ")
        visited.add(start)
        print(start, end=" ")
        for neighbor in sorted(self.graph[start]):
            if neighbor not in visited:
                self.dfs(neighbor, visited, False)
        if is_first:
            print()

    def bfs(self, start):
        if start not in self.graph:
            print("Vertex awal tidak ditemui.")
            return
        visited = set([start])
        queue = deque([start])
        print("Traversal BFS:", end=" ")
        while queue:
            vertex = queue.popleft()
            print(vertex, end=" ")
            for neighbor in sorted(self.graph[vertex]):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        print()

def main():
    g = Graph()
    while True:
        print("\nMENU:")
        print("1. Tambah Vertex")
        print("2. Hapus Vertex")
        print("3. Tambah Edge")
        print("4. Hapus Edge")
        print("5. Tampilkan graph (Matrix)")
        print("6. Traversal DFS")
        print("7. Traversal BFS")
        print("8. Quit")
        
        pilihan = input("Pilih menu (1-8): ")
        
        if pilihan == '1':
            v = input("Masukkan nama vertex: ")
            g.add_vertex(v)
        elif pilihan == '2':
            v = input("Masukkan nama vertex untuk dihapus: ")
            g.remove_vertex(v)
        elif pilihan == '3':
            v1 = input("Masukkan vertex 1: ")
            v2 = input("Masukkan vertex 2: ")
            g.add_edge(v1, v2)
        elif pilihan == '4':
            v1 = input("Masukkan vertex 1: ")
            v2 = input("Masukkan vertex 2: ")
            g.remove_edge(v1, v2)
        elif pilihan == '5':
            g.display_matrix()
        elif pilihan == '6':
            start = input("Masukkan vertex awal untuk DFS: ")
            g.dfs(start)
        elif pilihan == '7':
            start = input("Masukkan vertex awal untuk BFS: ")
            g.bfs(start)
        elif pilihan == '8':
            print("Program tamat.")
            break
        else:
            print("Pilihan tidak sah.")

if __name__ == "__main__":
    main()